[mono] Add support for UnmanagedCallersOnlyAttribute

lambdageek opened the pull request:
When we see a method tagged with `UnmanagedCallersOnlyAttribute` in an `ldtn` instruction, load a pointer to a native-to-managed wrapper instead of a pointer to the managed method.

See https://github.com/dotnet/csharplang/blob/master/proposals/functionpointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute

TODO:
- [x] Throw exceptions for all the unsupported scenarios:
   - Attempting to call by creating a delegate
   - Invoking via reflection
   - Using unsupported argument or return types
- [x] ~Figure out why the CoreCLR tests are either not running or somehow passing.~
   ~The tests are in https://github.com/dotnet/runtime/tree/master/src/coreclr/tests/src/Interop/UnmanagedCallersOnly which doesn't seem to be in https://github.com/dotnet/runtime/blob/master/src/coreclr/tests/issues.targets but in that case Mono should be failing a number of the tests in there on CI.~
   It's because we only run P0 tests on CI for Mono.  I temporarily set it to P0 for this PR just to watch the tests fail.
- [x] Revert tests back to P1

lambdageek committed:
[marshal] Add mono_method_has_unmanaged_callers_only_attribute

lambdageek committed:
[marshal] Allow calls to mono_marshal_get_managed_wrapped without a delegate
class

In that case, create a wrapper based on the signature of the method itself.

TODO:
 - check that the method is static
 - check that the method does not have any marshal info
   (currently it's assumed to not be present and ignored)

lambdageek committed:
[aot] Allow decode_method_ref to decode NATIVE_TO_MANAGED wrappers without a
delegate class

Bump the AOT file format

lambdageek committed:
[mini] return ptr to wrapper in mono_ldftn for UnmanagedCallersOnly methods

lambdageek committed:
[interp] ldftn will return a native-to-managed wrapper to UnmanagedCallersOnly methods

lambdageek committed:
[tests] Temporarily set UnmanagedCallersOnly tests to Pri0

vargaz commented on src/mono/mono/mini/aot-runtime.c:
```diff
@@ -1313,8 +1313,10 @@ decode_method_ref_with_target (MonoAotModule *module, MethodRef *ref, MonoMethod
 			if (!m)
 				return FALSE;
 			klass = decode_klass_ref (module, p, &p, error);
+#if 0
```
Would be better to explicitly encode a NULL class using an extra byte.

vargaz commented:
Looks ok.

lambdageek commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -6755,3 +6769,34 @@ get_marshal_cb (void)
+	ERROR_DECL (attr_error);
+	MonoClass *attr_klass = NULL;
+#ifdef ENABLE_NETCORE
+	attr_klass = = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```
```suggestion
	attr_klass = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```

lambdageek committed:
fixup build

lambdageek committed:
fixup null

lambdageek committed:
aot: emit byte when we don't expect a class

lambdageek commented on src/mono/mono/mini/jit-icalls.c:
```diff
@@ -39,6 +39,18 @@ mono_ldftn (MonoMethod *method)
 	gpointer addr;
 	ERROR_DECL (error);
 
+	if (G_UNLIKELY ((method->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL) == 0 &&
+			method->wrapper_type == MONO_WRAPPER_NONE &&
+			mono_method_has_unmanaged_callers_only_attribute (method))) {
```
There's a perf problem here
https://github.com/mono/mono/pull/20060#discussion_r449288732

It's actually worse for the JIT than for the interpreter (interp adds the wrapper at transform time, not at execution time; the JIT does this attribute lookup every time `mono_ldftn` gets called at execution time).

I think we can improve things a bit by checking for the attribute in `method-to-ir` and calling a slightly different icall that unconditionally returns the addr of a wrapper.

lambdageek committed:
jit: move wrapper creation to method-to-ir, not mono_ldftn

Do it at IR generation of the caller, not every time the ldftn is executed

lambdageek committed:
fix typo

lambdageek committed:
don't get callconv from Invoke if there's no invoke

alexischr commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -6755,3 +6770,34 @@ get_marshal_cb (void)
+gboolean
+mono_method_has_unmanaged_callers_only_attribute (MonoMethod *method)
+{
+	/* TODO: if this is slow and called repeatedly for the same method, add a bit to MonoMethod that is true if the
```
Just a reminder to remove this comment as you moved the invocation to method_to_ir() so it (probably) doesn't apply anymore

alexischr reviewed (commented):
LGTM, documenting that this will increase AOT binaries sizes by 1 byte per native-to-managed wrapper AFAICT

lambdageek committed:
jit: don't create a jump trampoline for ldftn of a native-to-managed wrapper

The wrapper might be called from a thread that's not attached to the runtime,
and the jump trampoline will look at TLS vars that are not initialized

lambdageek committed:
interp: transform LDFTN into LDC of a create_method_pointer for UnmanagedCallersOnly method

lambdageek commented:
> this will increase AOT binaries sizes by 2 bytes per native-to-managed wrapper AFAICT

1 byte, I think - AOT uses the same value encoding as CIL images - small values (<127) encode to one byte.

lambdageek committed:
remove obsolete comment

lambdageek commented:
Should pass the positive tests now.  Negative tests will still fail.

lambdageek commented:
I think preventing calls via `calli` will be difficult using Mono's current design - the native to managed wrapper can't really tell it's being called from a managed method.

lambdageek committed:
marshal: throw invalid program exception for instance and generic methods

lambdageek commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -11106,6 +11106,17 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+					mono_method_has_unmanaged_callers_only_attribute (cmethod))) {
+				MonoClass *delegate_klass = NULL;
+				MonoGCHandle target_handle = 0;
+				cmethod = mono_marshal_get_managed_wrapper (cmethod, delegate_klass, target_handle, cfg->error);
+				CHECK_CFG_ERROR;
```
This isn't quite right.  On Mono we will throw when JITing the method that has the `ldftn` instruction that has a disallowed `cmethod` (e.g. an instance method).  On CoreCLR they throw when the method runs.  So we would need to capture the error and emit a throw here.

And probably something similar in the interpreter

lambdageek committed:
Emit IPE throw instead of aborting JIT or interp compilation for bad UnmanagedCallersOnly methods

Instead of throwing while JITing (or transforming), throw when the LDFTN is
executed.

lambdageek committed:
Use domain mempool for IPE error message

lambdageek commented:
Some progress on the negative tests. The InvalidProgramException ones should be in good shape.

lambdageek committed:
disallow delegate constructor calls on UnmanagedCallersOnly methods

lambdageek commented:
The `TestUnmanagedCallersOnlyValid` test is failing with the interpreter due to https://github.com/dotnet/runtime/issues/38897

lambdageek committed:
throw IPE if UnmanagedCallersOnly method has non-blittable args

lambdageek commented:
JIT is passing all the tests except `TestPInvokeMarkedWithUnmanagedCallersOnly`.  Working on that one now.

lambdageek committed:
disallow DllImport and UnmanagedCallersOnly together

throw NotSupportedException

lambdageek commented:
Once I see the JIT pass all the tests, I'll bump the test back to Pri1 and then the PR is basically complete.

lambdageek commented:
From https://helix.dot.net/api/2019-06-17/jobs/51ac422f-058f-4fa3-aa34-f41bb363a4fb/workitems/Interop/console the "runtime (Mono Pri0 Runtime Tests Run OSX x64 release)" JIT job:
```
Interop.UnmanagedCallersOnly.XUnitWrapper           Total:  1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.252s
```

As previously mentioned, the interpreter job is failing due to #38897, so this is the best we'll get for now.

lambdageek committed:
Change UnamangedCallersOnlyTest.csproj back to Pri1

CoffeeFlux commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3940,10 +3960,55 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
+static gboolean
+type_is_blittable (MonoType *type)
+{
+	if (type->byref)
```
Maybe also check for `mono_type_is_reference (type)`?

lambdageek replied:
I think I want to look at the `MonoClass:blittable` bit for all of those.  This is just to get the `int&` sort of types.

CoffeeFlux reviewed (approved):
Lots of questions but looks fine to me. Would be really good if Zoltan gets a chance to look at this before it's merged since I'm not terribly confident in my review of the code generation parts. Thanks!

lambdageek commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3661,6 +3662,25 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
+		csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
+		csig->pinvoke = 0;
+		res = mono_mb_create_and_cache_full (cache, method, mb, csig,
+											 csig->param_count + 16, info, NULL);
```
I don't know. I'm cargo-culting the other `mono_mb_create_and_cache_full` calls in this function.

CoffeeFlux replied:
Ugh, that's what I feared. @vargaz any idea? For my own learning if nothing else :)

vargaz replied:
No idea, it probably comes from the fact that these are wrappers so the deepest stack will be calling the 'wrapped' method, so param_count + <const> is good for max stack.

lambdageek commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -6755,3 +6843,31 @@ get_marshal_cb (void)
+	ERROR_DECL (attr_error);
+	MonoClass *attr_klass = NULL;
+#ifdef ENABLE_NETCORE
+	attr_klass = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```
yes, if the attribute is not used anywhere, I think.  So we call the `_try_get_` function and return FALSE if it returns NULL.

CoffeeFlux replied:
Ah okay, you're using `_try_get_`. Should be good then.

lambdageek commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3940,10 +3960,55 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
+	case MONO_TYPE_R8:
+	case MONO_TYPE_I:
+	case MONO_TYPE_U:
+	case MONO_TYPE_CHAR:
```
I think char shouldn't be here. Doesn't look like it's supposed to be blittable. (`mono_class_setup_mono_type` doesn't set the blittable bit on it)

CoffeeFlux replied:
Ah yeah, good catch. Yes, it's definitely not for the same reason as strong—ambiguity over converting to ANSI, UTF-8, or UTF-16.

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Ryan Lucia <ryan@luciaonline.net>

lambdageek committed:
System.Char isn't blittable

lambdageek committed:
fixup non-blittable exception message

lambdageek committed:
non-netcore doesn't have UnmanagedCallersOnlyAttribute

lambdageek committed:
remove dead code

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3940,10 +3958,54 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
 #endif
 
+static gboolean
+type_is_blittable (MonoType *type)
```
There is already a class->blittable flag.

CoffeeFlux replied:
We discussed this, but it's not the same because of byref types.
